### PR TITLE
[BEAM-14053] [CdapIO] Add wrapper class for CDAP plugin

### PR DESCRIPTION
**Pull request by ktttnv**
**Warning:** This PR contains changes from [PR "[BEAM-14048] [CdapIO] Add ConfigWrapper for building CDAP PluginConfigs"](https://github.com/apache/beam/pull/17051) and should be merged after it

_Changes:_

- Added wrapper classes for source and sink CDAP plugins
- Added builder for plugin wrapper class
- Added possibility to provide Hadoop configuration

pluginWrapper3

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by chamikaramj** [comment]
Retest this please

File: sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/PluginConstants.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.cdap;
+
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.OutputFormat;
+
+public final class PluginConstants {
```
Comment (Amar3tto) [inline sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/PluginConstants.java:25]: Could you please provide description for this class and enums?

Comment (ktttnv) [reply sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/PluginConstants.java:28]: Done, thank you!

**Comment by Amar3tto** [comment]
Retest this please

**Comment by ktttnv** [comment]
R: @chamikaramj

**Comment by chamikaramj** [comment]
Could you clarify the diff that should be reviewed here ?
(since this contains code from https://github.com/apache/beam/pull/17051 and https://github.com/apache/beam/pull/17104)

**Comment by Amar3tto** [comment]
> Could you clarify the diff that should be reviewed here ? (since this contains code from #17051 and #17104)

Please review these files:
```
Plugin.java
PluginConstants.java
PluginTest.java
```

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
@chamikaramj this PR is ready for review

File: sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/Plugin.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+      return new BatchSourceContextImpl();
+    }
+  }
+
+  /** Gets value of a plugin type. */
+  public Boolean isUnbounded() {
+    Boolean isUnbounded = null;
+
+    for (Annotation annotation : getPluginClass().getDeclaredAnnotations()) {
+      if (annotation.annotationType().equals(io.cdap.cdap.api.annotation.Plugin.class)) {
```
[the remark this answers is not in the document]
Comment (Lizzfox) [reply sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/Plugin.java]: Thank you, that's helpful. Done

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
@chamikaramj PR is ready to be merged

**Pull request merged**